### PR TITLE
table: memoize TableCommon.getCols

### LINKED ISSUE #53604: TableCommon.getCols() can be memorized
## Enhancement
Every call of getCols() re-calculates the result, which is time-consuming. Since we assume the schema of the table won't change in the lifetime of a TableCommon struct, we can memoize the function.

### PR DESCRIPTION
**Pull request by ekexium**
### What problem does this PR solve?

Issue Number: close #53604 

Problem Summary:

`getCols` does duplicated calculation, which wastes CPU time.

### What changed and how does it work?
Memoize this function.

Proportion of time in the microbenchmark of AddRecord (with pipelined DML and ART memdb)

Previous:
image

This PR:
image

### Check List

Tests 

- [x] Unit test - micro benchmark + existing test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/table/tables/tables.go [hunk 1]
```diff
@@ -82,6 +82,10 @@ type TableCommon struct {
 	// recordPrefix and indexPrefix are generated using physicalTableID.
 	recordPrefix kv.Key
 	indexPrefix  kv.Key
+
+	// avoid duplicated calculations of Cols
+	// mode -> cols
+	getColsCache [][]*table.Column
```
Comment (Defined2014) [inline pkg/table/tables/tables.go:88]: Maybe we could use `hiddenColumns`, `visibleColumns` and `publicColumns` instead of this one.

Comment (ekexium) [reply pkg/table/tables/tables.go:88]: When are these fields supposed to be set (before this PR)? I don't find anywhere initializing them

Comment (Defined2014) [reply pkg/table/tables/tables.go:88]: Seems never be setted. /cc @zimulala

Comment (ekexium) [reply pkg/table/tables/tables.go:88]: Seems https://github.com/pingcap/tidb/pull/50062 introduced the regression. It removed the default initialization, but mistakenly making them never initialized.

File: pkg/table/tables/bench_test.go [hunk 2]
```diff
@@ -0,0 +1,39 @@
+package tables_test
```
Comment (hawkingrei) [inline pkg/table/tables/bench_test.go:1]: Please add a licenser header.

**Comment by ekexium** [comment]
Test failed because `IMPORT INTO` concurrently calls `AddRecord`, which calls `WritableConstraint()`, resulting in data race.
I'd rather let lightning manually initialize these fields than introducing something like sync.Once here.

File: pkg/table/tables/tables.go [hunk 3]
```diff
@@ -300,26 +315,26 @@ func (t *TableCommon) getCols(mode getColsMode) []*table.Column {
 
 // Cols implements table.Table Cols interface.
 func (t *TableCommon) Cols() []*table.Column {
-	if len(t.publicColumns) > 0 {
-		return t.publicColumns
+	if len(t.publicColumns) == 0 {
+		t.publicColumns = t.getCols(full)
```
Comment (zimulala) [inline pkg/table/tables/tables.go:319]: It seems that there is a parallel call problem here, such as multiple sessions using the same table in the same version of Infoschema.

**Review by Defined2014 — APPROVED** [review]

**Review by zimulala — APPROVED** [review]
LGTM

**Pull request merged**